### PR TITLE
Validation of  `fn` on `input_col` for Datapipes

### LINKED ISSUE #362: Add validation between `input_col` and `fn` for Callable DataPipes
### 🚀 The feature

Using `inspect` to validate the `input_col` matching `fn` input for `Mapper`, `Flatmapper` and other DataPipe using both `input_col` and `fn` as the arguments.
- [`Mapper`](https://github.com/pytorch/pytorch/blob/6dc71461e11122494a303214a72b12adb353dc4d/torch/utils/data/datapipes/iter/callable.py#L55-L57)
- [`Filter`](https://github.com/pytorch/pytorch/blob/ec591087fb95525b430f2bbcdd023567e2fb1cc9/torch/utils/data/datapipes/iter/selecting.py#L13-L47)
- [`Flatmap`](https://github.com/pytorch/data/blob/732307a1dcea8686bff9a855bcd67c32d2087f8e/torchdata/datapipes/iter/transform/callable.py#L93-L123)
- [`BatchMap`](https://github.com/pytorch/data/blob/732307a1dcea8686bff9a855bcd67c32d2087f8e/torchdata/datapipes/iter/transform/callable.py#L15-L59)
- etc

### Motivation, pitch

This would help users to identify the wrong arguments at the construction time for such `IterDataPipe` rather than at iteration time.

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by xiurobert**
Fixes #362 

### Changes

- Document input_col behavior on `FlatMapperIterDataPipe`
- Checks the arguments of mapping function against `input_col` in `FlatMapperIterDataPipe`

### CODE REVIEW

**Comment by xiurobert** [comment]
Oh, hold on. This will break test_flatmap_iterdatapipe

**Comment by xiurobert** [comment]
This breaks `test_map_batches_iterdatapipe`, specifically 683 (since there is only 1 argument)

**Comment by xiurobert** [comment]
@ejguan, should I modify `fn_2_cols` or is that not desired?

**Comment by ejguan** [comment]
Could you please open a PR in PyTorch core by creating the validation function first?
- [`Mapper`](https://github.com/pytorch/pytorch/blob/77b6885a225db0e4dfe331d3e98a5a456c672c2c/torch/utils/data/datapipes/iter/callable.py#L17)
- [`Filter`](https://github.com/pytorch/pytorch/blob/dd620c45752ab746bdd6bc47e8a2425a2f79a702/torch/utils/data/datapipes/iter/selecting.py#L14)
Then, TorchData can import such validation function from PyTorch Core and reuse it.

**Comment by xiurobert** [comment]
Oh, good idea. Any specific file I should put it in?

**Comment by ejguan** [comment]
You can put it in https://github.com/pytorch/pytorch/blob/master/torch/utils/data/datapipes/utils/common.py

**Comment by xiurobert** [comment]
Are the CI failures normal?

File: torchdata/datapipes/iter/transform/callable.py [hunk 1]
```diff
@@ -65,6 +96,7 @@ def __init__(
         assert batch_size > 0, "Batch size is required to be larger than 0!"
         self.batch_size = batch_size
         self.input_col = input_col
+        ensure_fn_works(fn, input_col)
```
Comment (ejguan) [inline torchdata/datapipes/iter/transform/callable.py:99]: Let's remove this part for now. I need to double check with the behavior of `map_batches`

**Review by ejguan — COMMENTED** [review]
The Error is related to `map_batches`. Let's remove the check for `map_batches` for now.

The logic of `ensure_fn_works` seems reasonable. Like I said before, we can move this function to PyTorch core to prevent duplicate implementations.

File: torchdata/datapipes/iter/transform/callable.py (hunk 1 above)
Comment (xiurobert) [reply torchdata/datapipes/iter/transform/callable.py:99]: I will take out both the function definition and the checks then until core merges my checks

**Comment by xiurobert** [comment]
If #293 goes through this function may need to be copied back over here

**Comment by xiurobert** [comment]
@ejguan Okay to continue working on this now that core has merged?

**Comment by ejguan** [comment]
> @ejguan Okay to continue working on this now that core has merged?

Yeah. You can do that. But, you might wait until my PR landed as your PR breaks a few tests for TorchVision and I fixed them in my PR.

**Comment by xiurobert** [comment]
Let me close and reopen this now that the validator function has been merged.

**Pull request closed without merging**